Add basic motion profiles

brendanburkhart opened the pull request:
Rough prototype, may not work correctly

brendanburkhart committed:
Add basic motion profiles

Rough prototype, may not work correctly

howardabrams commented on motion_profiles.py:
```diff
@@ -0,0 +1,101 @@
+         of (slope, intercept)) intersect
+        """
+        # Same slopes means parallel or collinear
+        if line1[0] == line2[0]:
```
Showing `line1[0]`looks _magical_. Perhaps we should assign these values to variables to be more descriptive? Even `x1` and `y1` imply a coordinate, and might be better.

howardabrams commented on motion_profiles.py:
```diff
@@ -0,0 +1,101 @@
+
+        ** Note: The units don't matter, as long as they are all the same - but anyone**
+        **caught using non-metric units will be publicly shamed.**
+        """
```
Nice comment!

howardabrams commented on motion_profiles.py:
```diff
@@ -0,0 +1,101 @@
+            This is equivalent to the area of a right-angled
+            triangle with base `time` and hypotenuse described by the line"""
+            return 0.5 * time * ((slope * time) + intercept)
+
```
Inner functions can be a great way to abstract complex features, however, this visually looks like a mistake 😉

Since we can't (I don't think) see the document associated with the `triangle_area` function, perhaps it would look better if it was a comment before it, and explains that it is an inner, helper function?

howardabrams reviewed (commented):
Perhaps we should include some simple unit tests to demonstrate this code? Otherwise, it looks like a great start!

brendanburkhart commented:
Yeah, once I've ironed out a few issues I will setup a test suite

brendanburkhart committed:
Improved motion profile variable names and comments to increase readability.

brendanburkhart committed:
Add tests for motion profile line intersection

Fix calculation of deceleration in MotionProfile.generate_motion_profile

calebeby commented:
@BrendanBurkhart we need to integrate this with the robotpy tests so that it is run when we do `py -3 robot.py test`

brendanburkhart commented:
Yah,  I have some tests written and the test command finds them quite nicely

calebeby commented:
Oh, didn't realize

calebeby commented:
Do we want to put all tests in the tests folder?

calebeby commented:
Oh, they already are. Nvm

brendanburkhart commented:
I was going to put a test module corresponding to each module we want to test in the test folder, but I haven't used py.test much so I don't really know what the best way to do it is

brendanburkhart committed:
Fix motion profile calculations to correctly handle triangular profiles

brendanburkhart committed:
Adds tests for motion profile generation

Includes both trapezoidal and triangular profiles.

calebeby commented on motion_profiles.py:
```diff
@@ -0,0 +1,107 @@
+    acceleration_time: float
+    deceleration_time: float
+    max_speed: float
+    end_time: float
```
What are these for? Aren't these variables used in method scope rather than class scope?

brendanburkhart committed:
Fixes motion profiles position calculations

Also adds more documentation to motion profile generation, rearrange fields of MotionProfile for clarity.

In MotionProfile tests, rearranged fields to match MotionProfile, removed unneeded intersection tests (intersection is no longer used).

brendanburkhart committed:
Adds motion profile rendering

Updates readme to document motion profile rendering.

brendanburkhart committed:
Improve motion profile documentation

Change how time constraints work in the motion profile position calculations so it can be used past the end time.

brendanburkhart committed:
Adds unit tests for motion profile position calculations

calebeby commented:
I think maybe we shouldn't merge this until we have integrated this with the `drivetrain` component.

brendanburkhart commented:
I agree it shouldn't be merged yet. How are we going to integrate it with the `drivetrain`? I was under the impression an autonomous would be reading out of the motion profile, into a PID and then setting the drivetrain, so the drivetrain doesn't know whether it is being controlled manually or by a profile. But I suppose the drivetrain could have some sort of `follow_profile` command...

howardabrams commented:
I actually could go either way. We _could_ merge it, but in such a way that it doesn't break anything (like `drivetrain`). Then we could integrate it with another commit.

calebeby commented:
Because it is extremely likely that we will have multiple autonomous modes, we shouldn't have the PID code in each auto file. I think that it is much better if we put the PID logic component-level, so that `drivetrain` has `drive_forward(meters=3)`, `turn_right(degrees=90)`, etc. Then our auto files will be extremely small and simple.

brendanburkhart commented:
I agree that bulk of the PID logic shouldn't go in the autonomous files. I see two options on where to put it, we could create some sort of `MotionProfileExecutor` that combines a motion profile and a PID, you would feed in the current position and time and receive the velocity/angles to drive at. The other is your example, verb commands attached to the drivetrain. I'm leaning towards your way, but it seems like that should be a separate piece of functionality and the `drivetrain` should just drive the wheels. On the other hand, we are already using the `wpilib.RobotDrive`, so they would be separate regardless.

howardabrams commented on motion_profiles.py:
```diff
@@ -0,0 +1,145 @@
+class MotionProfile(NamedTuple):
+    """Motion profile representing a specific desired move"""
+    """Robot's acceleration - slope of acceleration triangle"""
+    acceleration: float
```
How does this Python code work? I get errors each time I load this. Is the goal for these to be instance variables? Like:

    self.acceleration = 0.0

brendanburkhart replied:
I’m on my phone right now so I can’t help very much. What errors are you getting?

brendanburkhart replied:
The typing.NamedTuple module is a standard library component, it will turn them into instance variables on creation. It’s an immutable tuple whose fields can be accessed like instance members.

calebeby replied:
What errors?

BrendanBurkhart replied:
Perhaps he isn't running python >= 3.5, I believe the named tuple was only added then.

brendanburkhart commented:
@calebeby Should we merge this now and integrate with the `drivetrain` later?

calebeby commented:
I think that we should leave it out for now because it is unused code. I think that we should integrate `drivetrain` in this PR.

brendanburkhart committed:
Merge branch 'master' into bfb/add-motion-profiles

brendanburkhart committed:
Begin work on integrating motion profiles into drivetrain

brendanburkhart committed:
Implements mag encoder in physics.py

brendanburkhart committed:
Merge branch 'bfb/add-motion-profiles' of https://github.com/Pigmice2733/Bunnybot-2017 into bfb/add-motion-profiles

brendanburkhart committed:
MotionProfile handles negative distances correctly

Adds tests for negative distances, and includes negative profile in render_motion_profiles. render_motion_profiles outputs images into separate directory now.

brendanburkhart committed:
Add gyro channel correctly in physics.py

brendanburkhart committed:
Exract motion profile manipulation into an executor class in drivetrain

brendanburkhart committed:
Create example autonomous modes using motion profiles

brendanburkhart committed:
Adds 'backward' command to drivetrain

Also adds documentation for forward/backward/rotate commands.

brendanburkhart committed:
Improved gyro handling in drivetrain

Gyro reset will not interfere with other consumers of the gyro.

brendanburkhart committed:
Merge branch 'master' into bfb/add-motion-profiles

calebeby commented on components/drivetrain.py:
```diff
@@ -1,21 +1,85 @@
+        coefs = PIDCoefficients(p=1.0, i=0.0, d=0.0)
+        self.pid = PIDController(coefs, 1, -1)
+
+    def get_encoder_position(self):
```
What's this for?

brendanburkhart replied:
I think you're reviewing an outdated version. Try switching to the latest commit.

calebeby commented on autonomous/forward.py:
```diff
@@ -0,0 +1,19 @@
+        self.next_state('wait_till_completion')
+
+    @state
+    def wait_till_completion(self):
```
I'm not sure I like this. Could we do something like this instead?

```python
@state(first=True)
def forward(self):
    done = self.drivetrain.forward(feet=25)
    if done:
          self.next_state('whatever')
```
This means that `drivetrain.forward` would be called constantly, and would return a boolean, whether it is done. I think we should `update` the `ProfileExecutor` when `forward` and `rotate` are called, rather than in `execute`

brendanburkhart replied:
Hmm. So what should happen if `drivetrain.forward(feet=20)` had been called, and then `drivetrain.forward(feet=19)` was called?

calebeby replied:
It wouldn't override.

brendanburkhart replied:
So after the first call to `drivetrain.forward(feet=_)`, the parameters are just ignored?

calebeby replied:
Actually, we could keep the goal distance (or angle) on self, and it would override if it is different than the current goal distance

brendanburkhart replied:
Ok. I can live with that. Since there are multiple distance parameters to `drivetrain.forward`, does wrapping them in a tuple and storing them that way seem good? Alternatively, we *could* just ignore the arguments and force a call to `drivetrain.cancel_motion_profile()`.

calebeby replied:
Just combine them all to metric and save the metric value.

brendanburkhart replied:
Rather then using `drivetrain.cancel_motion_profile` to switch profiles?

brendanburkhart replied:
I was thinking, we really shouldn't ever have to switch profiles, so storing the arguments in expectation of that seems a bit weird. Also it screws with expected behavior, as calling `drivetrain.forward` will sometimes override, sometimes not. As another option (I'm really just throwing them out so we consider everything), there could be some `override` parameter specifying whether you want to override existing profiles.

calebeby replied:
Yeah, I guess cancelling might be a better choice here. Maybe (in this PR or later on) error or warn if the parameters are changed without cancelling?

brendanburkhart replied:
Ok, sounds good. I think warning is a good idea.

brendanburkhart committed:
Improve Drivetrain motion profile interface

calebeby commented on components/drivetrain.py:
```diff
@@ -1,21 +1,180 @@
+    def _update_executor(self):
+        executor_finished = self.profile_executor.update()
+        if executor_finished:
+            return True
```
Can we cancel the executor here?

brendanburkhart replied:
That’s what I did originally, but it forces the consumer to stop calling forward or rotate immediately or a new profile will be started. Should I cancel there anyway?

brendanburkhart committed:
Change how profiles are finished

calebeby commented on components/drivetrain.py:
```diff
@@ -1,21 +1,199 @@
+
+        Returns `False` while executing, `True` once done. Continuing
+         to update when done will start new profile."""
+        self.forward(-feet, -inches, -meters)
```
You need to `return` here, right?

brendanburkhart replied:
Yep, thanks

brendanburkhart committed:
Fix returns in drivetrain commands

brendanburkhart committed:
Remove unused ProfileType from drivetrain.

brendanburkhart committed:
Remove unused import from drivetrain.py

howardabrams commented on autonomous/forward.py:
```diff
@@ -0,0 +1,14 @@
+
+class Forward(AutonomousStateMachine):
+    MODE_NAME = "Forward"
+    DEFAULT = True
```
So the constants won instead of the the Enums, eh?

brendanburkhart replied:
Robotpy needs those for the autonomous selector, I'm still somewhat undecided about the rest. Constants just don't tell you if they belong in groups, but that can remedied with lots of comments, so I think we'll probably be using constants for now.

calebeby commented:
Thanks for doing this Brendan, good job. I'd like Alex or Ben to review this, because I think we need fresh eyes.

BrendanBurkhart commented:
Yeah, I think this pr got a bit too big. I want Howard's review especially, but he didn't seem to make it through a review last night.

howardabrams committed:
Merge branch 'master' into bfb/add-motion-profiles
